javascript/code-challenges/challenges-10.test.js

```diff
@@ -0,0 +1,364 @@
+'use strict';
+
+/* ------------------------------------------------------------------------------------------------
+CHALLENGE 1 - Review
+
+Write a function named returnTen, takes in a string and uses split and splice to return the last 10 characters from that string as elements of an array.
+
+------------------------------------------------------------------------------------------------ */
+
+function returnTen(str){
+  return str.split('').splice(str.length - 10, 10);
+}
+
+// Why can you not delete backwards as well??
+
+/* ------------------------------------------------------------------------------------------------
+CHALLENGE 2
+
+Write a function named findMax that takes in a matrix of positive numbers and returns the number with the highest value.
+
+For example:
+[
+  [1, 3, 4, 5],
+  [4, 5, 6],
+  [23, 5, 5]
+]
+
+return: 23
+------------------------------------------------------------------------------------------------ */
+const findMax = (matrix) => {
+  let newArr = [];
+  for(let i = 0; i < matrix.length; i++){
+    for(let j = 0; j < matrix[i].length; j++){
+      newArr.push(matrix[i][j]);
+    }
+  }
+  return newArr.reduce((prev, curr) => prev > curr ? prev : curr);
+};
+
+/* ------------------------------------------------------------------------------------------------
+CHALLENGE 3
+
+Write a function named totalSum that takes in a matrix of numbers and returns the totalSum of all the numbers.
+
+For example:
+[
+  [1, 3, 4, 5],
+  [4, 5, 1],
+  [2, 5, 5]
+]
+
+return: 35
+------------------------------------------------------------------------------------------------ */
+const totalSum = (matrix) => {
+  let newArr = [];
+  for(let i = 0; i < matrix.length; i++){
+    for(let j = 0; j < matrix[i].length; j++){
+      newArr.push(matrix[i][j]);
+    }
+  }
+  return newArr.reduce((prev, curr) => prev + curr, 0);
```

Comment:
So, there's are better ways to go about this. You want to avoid nested loops if you can. They're more often than not one of the least efficient ways of writing code (you'll get into this in 401). Plus they're just confusing. There's a few ways to get around this: Multiple reduce statements, array.concat etc